src/component.d.ts

```diff
@@ -0,0 +1,120 @@
+import {
+    AllowedComponentProps,
+    ExtractPropTypes,
+    ComponentCustomProps,
+    ComponentOptionsMixin,
+    DefineComponent,
+    PropType,
+    VNodeProps,
+} from "vue";
+import Flatpickr from "flatpickr";
+
+type FlatpickrEvent = "onChange" | "onClose" | "onDayCreate" | "onDestroy" | "onKeyDown" | "onMonthChange" | "onOpen" | "onParseConfig" | "onReady" | "onValueUpdate" | "onYearChange" | "onPreCalendarPosition";
+
+declare const _default: DefineComponent<{
+    modelValue: {
+        type: PropType<string | number | any[] | Date>;
+        default: any;
+        required: true;
+        validator(value: unknown): boolean;
+    };
+    config: {
+        type: PropType<Flatpickr.Options.Options>;
+        default: () => {
+            readonly wrap: false;
+            readonly defaultDate: any;
+        };
+    };
+    events: {
+        type: PropType<FlatpickrEvent[]>;
+        default: () => ["onChange", "onClose", "onDestroy", "onMonthChange", "onOpen", "onYearChange"];
+    };
+    disabled: {
+        type: BooleanConstructor;
+        default: boolean;
+    };
+}, unknown, {
+    /**
+     * The flatpickr instance
+     */
+    fp: Flatpickr.Instance;
+}, {}, {
+    /**
+     * Get the HTML node where flatpickr to be attached
+     * Bind on parent element if wrap is true
+     */
+    getElem(): Element | null;
+    /**
+     * Watch for value changed by date-picker itself and notify parent component
+     *
+     * @param event
+     */
+    onInput(event: InputEvent): void;
+    /**
+     * @return HTMLElement
+     */
+    fpInput(): HTMLInputElement;
+    /**
+     * Blur event is required by many validation libraries
+     *
+     * @param event
+     */
+    onBlur(event: FocusEvent): void;
+    /**
+     * Flatpickr does not emit input event in some cases
+     */
+    onClose(selectedDates: Date[], dateStr: string): void;
+    /**
+     * Watch for the disabled property and sets the value to the real input.
+     *
+     * @param newState
+     */
+    watchDisabled(newState: boolean): void;
+}, ComponentOptionsMixin, ComponentOptionsMixin, ["blur", "update:modelValue", "on-change", "on-close", "on-destroy", "on-month-change", "on-open", "on-year-change", "on-value-update", "on-day-create", "on-parse-config", "on-ready", "on-pre-calendar-position", "on-key-down"], "blur" | "update:modelValue" | "on-change" | "on-close" | "on-destroy" | "on-month-change" | "on-open" | "on-year-change" | "on-value-update" | "on-day-create" | "on-parse-config" | "on-ready" | "on-pre-calendar-position" | "on-key-down", VNodeProps & AllowedComponentProps & ComponentCustomProps, Readonly<ExtractPropTypes<{
+    modelValue: {
```

Comment:
Why do we have to repeat the props definition ?

Reply:
Honestly... not quite sure. The typings for Vue are pretty rough, as you said, so it's easier to write a component in ts, then generate the types and go from there. So this is auto-generated then hand edited. I didn't mess with this part. We could test removing it and see if that causes any problems.